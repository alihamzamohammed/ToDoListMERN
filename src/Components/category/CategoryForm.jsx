import React, { useState, useEffect } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { Form, Button, Row } from 'react-bootstrap';

const CategoryForm = () => {

    const { paramId } = useParams();
    const [id, setId] = useState();
    const history = useHistory();
    const [title, setTitle] = useState("");
    const [response, setResponse] = useState("");
    const [disabled, setDisabled] = useState(false);

    const createCategory = async () => {
        await fetch(`http://localhost:5050/category/create`, {
            method: "POST",
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                name: title
            })
        });
        setResponse("Category was created successfully!");
    };

    const updateCategory = async () => {
        await fetch(`http://localhost:5050/category/update/${id}`, {
            method: "PUT",
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                name: title
            })
        });
        setResponse("Category was updated successfully!");
    };

    const deleteCategory = async () => {
        await fetch(`http://localhost:5050/category/delete/${id}`, {
            method: "DELETE"
        });
        setResponse("Category was deleted successfully!");
        setDisabled(true);
    };

    const create = (typeof paramId === "undefined" || paramId === null);

    useEffect(() => {
        async function readCategory() {
            const read = await fetch(`http://localhost:5050/category/read/${id}`);
            const json = await read.json()
            setTitle(json.name)
        }
        setId(paramId);
        if (!create) {
            readCategory()
        }
    }, [create, id, paramId])


    return (
        id === "0" ? <Row className="justify-content-md-center"><p className="">This category cannot be edited</p></Row> :
            create ?
                <>
                    <h1 className="display-3 title">Create Category:</h1>
                    <Form className="text-input">
                        <div className="mb-3">
                            <Form.Group>
                                <Form.Label htmlFor="title">Title</Form.Label>
                                <Form.Control type="text" placeholder="Enter title" name="title" value={title} onChange={event => setTitle(event.target.value)}></Form.Control>
                            </Form.Group>
                        </div>
                        <div className="mb-3 gx-5">
                            <Button variant="success" onClick={() => createCategory()} className="form-button">Create</Button>
                            <Button variant="primary" onClick={() => setTitle("")} className="form-button">Reset</Button>
                            <Button variant="warning" onClick={() => history.push("/")} className="form-button">Discard</Button>
                        </div>
                    </Form>
                    <div>
                        <p className="response">{response}</p>
                    </div>
                </>
                :
                <>
                    <h1 className="display-3 title">Edit Category:</h1>
                    <Form className="text-input">
                        <div className="mb-3">
                            <p className="text-muted">ID: {id}</p>
                        </div>
                        <div className="mb-3">
                            <Form.Group>
                                <Form.Label htmlFor="title">Title</Form.Label>
                                <Form.Control type="text" placeholder="Enter title" name="title" value={title} onChange={event => setTitle(event.target.value)} disabled={disabled}></Form.Control>
                            </Form.Group>
                        </div>
                        <div className="mb-3 gx-5">
                            <Button variant="success" onClick={() => updateCategory()} className="form-button" disabled={disabled}>Update</Button>
                            <Button variant="primary" onClick={() => setTitle("")} className="form-button" disabled={disabled}>Reset</Button>
                            <Button variant="warning" onClick={() => history.push("/")} className="form-button" disabled={disabled}>Discard</Button>
                            <Button variant="danger" onClick={() => deleteCategory()} className="form-button" disabled={disabled}>Delete</Button>
                        </div>
                    </Form>
                    <div>
                        <p className="response">{response}</p>
                    </div>
                </>
    )
}

export default CategoryForm;