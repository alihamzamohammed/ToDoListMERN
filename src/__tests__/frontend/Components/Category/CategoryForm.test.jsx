import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import "whatwg-fetch";

import { MemoryRouter as Router, Route } from "react-router-dom";
import CategoryForm from "../../../../Components/Category/CategoryForm";

describe("Category Form Tests", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    jest.resetAllMocks();
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  afterAll(() => {
    global.fetch.mockClear();
    delete global.fetch;
  });

  it("Renders the Edit Category Form with ID 0", () => {
    jest.mock("react-router", () => ({
      ...jest.requireActual("react-router"), // use actual for all non-hook parts
      useParams: () => ({
        paramId: "0",
      }),
    }));

    act(() => {
      render(
        <Router initialEntries={["/edit/category/0"]}>
          <Route path="/edit/category/:paramId">
            <CategoryForm></CategoryForm>
          </Route>
        </Router>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });

  it("Renders the Edit Category Form", async () => {
    const testCat = {
      name: "Test Category",
      _id: "123",
      dateAdded: "2021-07-06T14:20:56.175Z",
    };

    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(testCat),
      })
    );

    jest.mock("react-router", () => ({
      ...jest.requireActual("react-router"), // use actual for all non-hook parts
      useParams: () => ({
        paramId: "123",
      }),
    }));

    await act(async () => {
      render(
        <Router initialEntries={["/edit/category/123"]}>
          <Route path="/edit/category/:paramId">
            <CategoryForm></CategoryForm>
          </Route>
        </Router>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });

  it("Renders the Create Category Form", () => {
    act(() => {
      render(
        <Router location="/create/category">
          <CategoryForm></CategoryForm>
        </Router>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });
});
