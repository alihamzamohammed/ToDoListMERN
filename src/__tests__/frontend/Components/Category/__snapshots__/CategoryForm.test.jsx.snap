// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category Form Tests Renders the Create Category Form 1`] = `
<div>
  <h1
    class="display-3 title"
  >
    Create Category:
  </h1>
  <form
    class="text-input"
  >
    <div
      class="mb-3"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="title"
        >
          Title
        </label>
        <input
          class="form-control"
          name="title"
          placeholder="Enter title"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="mb-3 gx-5"
    >
      <button
        class="form-button btn btn-success"
        type="button"
      >
        Create
      </button>
      <button
        class="form-button btn btn-primary"
        type="button"
      >
        Reset
      </button>
      <button
        class="form-button btn btn-warning"
        type="button"
      >
        Discard
      </button>
    </div>
  </form>
  <div>
    <p
      class="response"
    />
  </div>
</div>
`;

exports[`Category Form Tests Renders the Edit Category Form 1`] = `
<div>
  <h1
    class="display-3 title"
  >
    Edit Category:
  </h1>
  <form
    class="text-input"
  >
    <div
      class="mb-3"
    >
      <p
        class="text-muted"
      >
        ID: 
        123
      </p>
      <p
        class="text-muted"
      >
        Date Added: 
        Tue, 06 Jul 2021 14:20:56 GMT
      </p>
    </div>
    <div
      class="mb-3"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          for="title"
        >
          Title
        </label>
        <input
          class="form-control"
          name="title"
          placeholder="Enter title"
          type="text"
          value="Test Category"
        />
      </div>
    </div>
    <div
      class="mb-3 gx-5"
    >
      <button
        class="form-button btn btn-success"
        type="button"
      >
        Update
      </button>
      <button
        class="form-button btn btn-primary"
        type="button"
      >
        Reset
      </button>
      <button
        class="form-button btn btn-warning"
        type="button"
      >
        Discard
      </button>
      <button
        class="form-button btn btn-danger"
        type="button"
      >
        Delete
      </button>
    </div>
  </form>
  <div>
    <p
      class="response"
    />
  </div>
</div>
`;

exports[`Category Form Tests Renders the Edit Category Form with ID 0 1`] = `
<div>
  <div
    class="justify-content-md-center row"
  >
    <p
      class=""
    >
      This category cannot be edited
    </p>
  </div>
</div>
`;
